ci/get_commit_message.py

```diff
@@ -0,0 +1,116 @@
+
+
+def get_pragma_values(commit_message, pragma):
+    """Get the commit pragma values.
+
+    Supports multiple pragma entries in the commit message.
+
+    Args:
+        commit_message (str): the commit message
+        pragma (str): the commit pragma name (case-sensitive)
+
+    Returns:
+        str: the values defined for the specified pragma
+    """
+    regex = rf'^{pragma}:\s*(.*)$'
+    return ' '.join(re.findall(regex, commit_message, flags=re.MULTILINE | re.IGNORECASE))
+
+
+def update_commit_pragma(commit_message, pragma, new_values):
+    """Update the existing commit pragma with new values.
+
+    Inserts the updated commit pragmas after the commit message title.
+
+    Args:
+        commit_message (str): the original commit message
+        pragma (str): the commit pragma name
+        new_values (str): values to assign for the new commit pragma. If None the old commit pragma
+            is simply removed.
+
+    Returns:
+        str: the updated commit message
+    """
+    regex = rf'^{pragma}:.*$'
+
+    # Remove the old commit pragma
+    commit_message = re.sub(regex, '', commit_message, flags=re.MULTILINE | re.IGNORECASE)
+    if new_values is None:
+        return commit_message
+
+    # Insert the new commit pragma
+    commit_message_split = commit_message.splitlines()
+    commit_message_split.insert(1, '')
+    commit_message_split.insert(2, f'{pragma}: {new_values}')
+    commit_message_split.insert(3, '')
+    return os.linesep.join(commit_message_split)
+
+
+def modify_commit_message_pragmas(commit_message):
+    """Modify the commit message pragmas.
+
+    Args:
+        commit_message (str): the original commit message
+
+    Returns:
+        str: the modified commit message
+    """
+    test_tags = set()
+    test_tags_md_on_ssd = 'pr,md_on_ssd'
+    test_tag_values = get_pragma_values(commit_message, 'Test-tag')
+    if test_tag_values:
+        # When 'Test-tags:' exists in the commit message use the same tags for the MD on SSD testing
+        test_tags.update(filter(None, test_tag_values.split(' ')))
+        test_tags_md_on_ssd = test_tag_values
+    else:
+        # Default to using the 'pr' tag
+        test_tags.add('pr')
+
+    feature_values = get_pragma_values(commit_message, 'Features')
+    if feature_values:
+        # Add daily and weekly feature tests when 'Features' exists
+        for feature in feature_values.split(' '):
+            test_tags.update(f'daily_regression,{feature}')
+            test_tags.update(f'full_regression,{feature}')
+
+    # Replace 'Features:' with 'Test-tags:'
+    commit_message = update_commit_pragma(commit_message, 'Features', None)
+    commit_message = update_commit_pragma(commit_message, 'Test-tag', test_tags)
+
+    if 'Test-tag-hw-medium-md-on-ssd' not in commit_message:
+        # Add tags for MD on SSD testing
+        commit_message = update_commit_pragma(
+            commit_message, 'Test-tag-hw-medium-md-on-ssd', test_tags_md_on_ssd)
```

Comment:
If I'm understanding correctly, this is adding a `Test-tag-hw-medium-md-on-ssd: ` commit pragma, defaulting to a value of `pr,md_on_ssd` to every PR of or landing to `master`, yes?

Where/how is this pragma consumed?

Reply:
Almost.  The issue is the intent of the `Functional Hardware Medium MD on SSD` stage was for it to typically run a subset of tests that take ~1hr to run.  To achieve this we tagged a specific set of tests (most notably 24 of 28 `daos_test/suite.py` tests) with the `md_on_ssd` tag and set the default tags for the stage to `pr,md_on_ssd` in the Jenkinsfile.  

This approach was intentionally used instead of using the `md_on_ssd` as a 'sizing' tag - like `hw,medium` or `provider` - so that when a test fix/change needed to be verified with both PMEM and MD on SSD it could be done in a single PR build by running the same tag in the `Functional Hardware Medium` (or `Functional Hardware Medium Verbs Provider` for `daos_test/suite.py` changes) and `Functional Hardware Medium MD on SSD` stages by using the `Test-tag:` pragma.

The unintended consequence of supporting this is when the `Features:` commit pragma is used we end up with a `Functional Hardware Medium MD on SSD` stage that is not only running the same test workload as the `Functional Hardware Medium` stage but also the `Functional Hardware Medium Verbs Provider` stage without the benefit of parallel execution.

So, the comprise being implemented here is to run the `Functional Hardware Medium MD on SSD` stage with `pr,md_on_ssd`, unless the `Test-tag:` commit pragma is used.


The `Test-tag-hw-medium-md-on-ssd` pragma is consumed by the `Functional Hardware Medium MD on SSD` stage through https://github.com/daos-stack/daos/blob/master/Jenkinsfile#L1141 and https://github.com/daos-stack/pipeline-lib/blob/master/vars/getFunctionalTestStage.groovy#L48-L49

Reply:
Updated behavior being implemented in this ticket: https://daosio.atlassian.net/browse/DAOS-14693?focusedCommentId=122358